### ISSUE TITLE
High level API for CoolProp

### ISSUE THREAD

**Issue by ibell** [CoolProp/CoolProp-museum#5]
@jowr, @squoilin etc.

I have fixed and added almost all the flash routines for pure and pseudo-pure fluids - see the v5 branch.  The first set of rough consistency tests all pass for REFPROP.  CoolProp still has a few cases that fail, and I need to add a few more flashes yet.  I'm currently revisiting the high-level interface, and I don't know what the best way to handle the high level interface is.  A couple of options:

``` C++
double r0 = PropsSI("Dmass","T",298,"P",101325,"REFPROP-Propane[0.5],Ethane[0.5]");
std::vector<double> z(2,0.5);
double r1 = PropsSI("Dmass","T",298,"P",101325,"REFPROP-Propane,Ethane", z);
// and/or
double r2 = PropsSIZ("Dmass","T",298,"P",101325,"REFPROP-Propane,Ethane", z);
```

Basically the options are some combination of
a) Encode the mole fractions in the string (definitely will be supported), need to decide on the delimiters though.  Perhaps we just stick with what we did before.  I like '-' between backend and fluids, but some fluids ('n-Propane', etc.) have a '-' in the name.  Or we could use "::" to get something like "REFPROP::Water".  Would be kind of like C++, which has some appeal.  I am less sure about the delimiters for mole fractions.
b) Overload (?) the PropsSI function with a form where you pass molar(!) fractions as the last parameter
c) Another function name entirely that takes mole fractions

**Comment by jowr** [comment]
Nice work Ian!

I think overloading an interface is not really a good idea. There are so
many sources of confusion.

C++ style delimiters sound great. Avoiding problems with fluids like
n-Pentane is a plus.

**Comment by ibell** [comment]
Agreed.  Let's discard the idea of an overload.  I think its only a good idea to overload functions with the same input variables.  In that case, it'll be a new function, called ??  What do you think about PropsSIZ? Does that work?  And what do we do about the DLL wrapper?  We will need to pass a double\* and an N. Should we overload?  So we have the prototypes

``` C++
// For C++
double PropsSIZ(string, string, double, string, double, string, vector<double>)
// For DLL
double PropsSIZ(string, string, double, string, double, string, double*, int)
```

That would be ok I think?

**Comment by ibell** [comment]
Or is it better to use `PropsSIX`?

Also, what do we do about brines (or other things) that would want mass fractions instead of mole fractions?  How do we make that clear?

**Comment by ibell** [comment]
@jowr So this is more or less what I have settled on I think.  They all yield the same results

``` C++
// First type (slowest, most string processing, exposed in DLL)
double r0A = PropsSI("Dmolar","T",298,"P",1e5,"Propane[0.5]&Ethane[0.5]"); // Default backend is HEOS
double r0B = PropsSI("Dmolar","T",298,"P",1e5,"HEOS::Propane[0.5]&Ethane[0.5]");
double r0C = PropsSI("Dmolar","T",298,"P",1e5,"REFPROP::Propane[0.5]&Ethane[0.5]");

std::vector<double> z(2,0.5);
// Second type (C++ only, a bit faster)
double r1A = PropsSI("Dmolar","T",298,"P",1e5,"Propane&Ethane", z); // Default backend is HEOS
double r1B = PropsSI("Dmolar","T",298,"P",1e5,"HEOS::Propane&Ethane", z);
double r1C = PropsSI("Dmolar","T",298,"P",1e5,"REFPROP::Propane&Ethane", z);

const double *pz = &(z[0]);
int n = z.size();
// Third type (for the DLL)
double r2A = PropsSIZ("Dmolar","T",298,"P",1e5,"Propane&Ethane", pz, n); // Default backend is HEOS
double r2B = PropsSIZ("Dmolar","T",298,"P",1e5,"HEOS::Propane&Ethane", pz, n);
double r2C = PropsSIZ("Dmolar","T",298,"P",1e5,"REFPROP::Propane&Ethane", pz, n);
```

**Comment by jowr** [comment]
Yes, I think that is fine. We might want to add a function like:

```
std::vector<double> z(2,0.5);
std::vector<std::string> names;
names.clear();
names.push_back(std::string("Propane"));
names.push_back(std::string("Ethane"));
// Fourth type?
double r4A = PropsSI("Dmolar","T",298,"P",1e5,names,z); // Default backend is HEOS
double r4B = PropsSI("Dmolar","T",298,"P",1e5,names,z,"HEOS");
double r4C = PropsSI("Dmolar","T",298,"P",1e5,names,z,"REFPROP");
```

or are that too many options?

**Comment by ibell** [comment]
Yes, that is a good idea, though I might move the backend before the fluids just for visual consistency.  This form would be more computationally efficient than encoding the names and backend all in one string.  I was headed towards this idea, but never quite made it.

**Comment by jowr** [comment]
Yes, but one nice thing about having the backend at the end is that the overload and the normal function call look much more alike... Well, no preferences here.

**Comment by ibell** [comment]
Hmm ok, I do see that point.  In the end, the types are different enough
that there is no chance of putting them in the wrong order, it won't
compile if you are in C++.  Let's maybe talk about this, but I think either
way is fine.
